GopherCAN.c

```diff
@@ -508,44 +490,38 @@ static S8 decode_parameter(CAN_INFO_STRUCT* param, U8* data, U8 start, U8 length
     }
 
     // restore original type
-    // TODO there is some floating point BS here when running the scale and offset on
-    // 32bit (and likely 64bit) numbers
     switch (param->TYPE) {
         case UNSIGNED8:
-            value = (value * param->SCALE) + param->OFFSET;
-            ((U8_CAN_STRUCT*) param)->data = (U8) value;
-            break;
-        case UNSIGNED16:
-            value = (value * param->SCALE) + param->OFFSET;
-            ((U16_CAN_STRUCT*) param)->data = (U16) value;
-            break;
-        case UNSIGNED32:
-            value = (value * param->SCALE) + param->OFFSET;
-            ((U32_CAN_STRUCT*) param)->data = (U32) value;
-            break;
-        case UNSIGNED64:
-            value = (value * param->SCALE) + param->OFFSET;
-            ((U64_CAN_STRUCT*) param)->data = (U64) value;
+            ((U8_CAN_STRUCT*)param)->data = (value * (U8)param->SCALE) + (U8)param->OFFSET;
             break;
         case SIGNED8:
-            value = (value * param->SCALE) + param->OFFSET;
-            ((S8_CAN_STRUCT*) param)->data = (S8) value;
+            ((S8_CAN_STRUCT*)param)->data = (value * (S8)param->SCALE) + (S8)param->OFFSET;
+            break;
+        case UNSIGNED16:
+            ((U16_CAN_STRUCT*)param)->data = (value * (U16)param->SCALE) + (U16)param->OFFSET;
             break;
         case SIGNED16:
-            value = (value * param->SCALE) + param->OFFSET;
-            ((S16_CAN_STRUCT*) param)->data = (S16) value;
+            ((S16_CAN_STRUCT*)param)->data = (value * (S16)param->SCALE) + (S16)param->OFFSET;
+            break;
+        case UNSIGNED32:
+            ((U32_CAN_STRUCT*)param)->data = (value * (U32)param->SCALE) + (U32)param->OFFSET;
             break;
         case SIGNED32:
-            value = (value * param->SCALE) + param->OFFSET;
-            ((S32_CAN_STRUCT*) param)->data = (S32) value;
+            ((S32_CAN_STRUCT*)param)->data = (value * (S32)param->SCALE) + (S32)param->OFFSET;
+            break;
+        case UNSIGNED64:
+            ((U64_CAN_STRUCT*)param)->data = (value * (U64)param->SCALE) + (U64)param->OFFSET;
             break;
         case SIGNED64:
-            value = (value * param->SCALE) + param->OFFSET;
-            ((S64_CAN_STRUCT*) param)->data = (S64) value;
+            ((S64_CAN_STRUCT*)param)->data = (value * (S64)param->SCALE) + (S64)param->OFFSET;
             break;
         case FLOATING:
-            value_fl = ((float) value * param->SCALE) + param->OFFSET;
-            ((FLOAT_CAN_STRUCT*) param)->data = value_fl;
+            // floats are signed values in data frame
+            if (length == 1) value_fl = (S8)value;
+            else if (length == 2) value_fl = (S16)value;
+            else if (length == 4) value_fl = (S32)value;
+            else if (length == 8) value_fl = (S64)value;
+            ((FLOAT_CAN_STRUCT*)param)->data = (value_fl * param->SCALE) + param->OFFSET;
```

Comment:
I think there should be one more 'else' that just decodes it as a U64 like before just in case something is not configured correctly